### DISCUSSION TITLE
Missing Development Dependency Indicators for GitHub Registry Response for Ruby Packages

### DISCUSSION THREAD

**Discussion by nbibler** [community/community#43861] [Actions] [answered]
### Select Topic Area

Bug

### Body

We're using GitHub Packages and Registries to host, in this case, private Ruby gem libraries for an Organization. Sometime in the past month (maybe shorter, like week or two?, honestly could be as recent as this weekend), I noticed that Dependabot started reporting unusual dependency update PR changes. Specifically: the requested changes via Dependabot started inexplicably including _development_ dependencies of dependent libraries. **Note:** I don't believe this is a Dependabot issue, I believe the Registry server's responses are incorrect and Dependabot is merely acting correctly on incorrect information received from it.

Investigating, I found that the `gem` responses do not appear to be correct from the GitHub Registry server (`rubygems.pkg.github.com`). Responses seem to fail to indicate that development dependencies are _development_ dependencies. Rather, it appears to be incorrectly reporting _all_ dependencies as production or runtime dependencies.

Here is a sample, correct dependency output for `rubygems.org` for a gem which includes both runtime and development dependencies:

```
% gem dependency -r dry-types -v 1.7.0
Gem dry-types-1.7.0
  bundler (>= 0, development)
  concurrent-ruby (~> 1.0)
  dry-core (~> 1.0, < 2)
  dry-inflector (~> 1.0, < 2)
  dry-logic (>= 1.4, < 2)
  rake (>= 0, development)
  rspec (>= 0, development)
  yard (>= 0, development)
  zeitwerk (~> 2.6)
```

Note that Rubygems recognizes **development** dependencies in that response.

Making a similar request for a GitHub Registry-hosted gem which contains both runtime and development dependencies:

```
% gem dependency -r private-gem-name1 -v 3.3.0 -s
"https://x-access-token:ghp_XXXXX@rubygems.pkg.github.com/myorg"
Gem private-gem-name1-3.3.0
  addressable (~> 2.3)
  private-gem-name2 (~> 2.1)
  appraisal (~> 2.0)
  bundler (~> 2.0)
  faraday (~> 2.0)
  license_finder (~> 7.0)
  omniauth (>= 1.1, < 3)
  rack-test (~> 2.0)
  rake (~> 13.0)
  redcarpet (~> 3.1)
  rspec (~> 3.0)
  rubocop-rails_config (~> 1.8)
  simplecov (~> 0.8)
  webmock (~> 3.0)
  yard-tomdoc (~> 0.7)
```

Note that in that response, there are **zero** dependencies listed as being "development" dependencies even though all but three of them are intended to be development-only dependencies (only 3 runtime dependencies):

```ruby
  spec.add_runtime_dependency 'addressable', '~> 2.3'
  spec.add_runtime_dependency 'private-gem-name2', '~> 2.1'
  spec.add_runtime_dependency 'omniauth', '>= 1.1', '< 3'

  spec.add_development_dependency 'appraisal', '~> 2.0'
  spec.add_development_dependency 'bundler', '~> 2.0'
  spec.add_development_dependency 'faraday', '~> 2.0'
  spec.add_development_dependency 'license_finder', '~> 7.0'
  spec.add_development_dependency 'rack-test', '~> 2.0'
  spec.add_development_dependency 'redcarpet', '~> 3.1'
  spec.add_development_dependency 'rake', '~> 13.0'
  spec.add_development_dependency 'rspec', '~> 3.0'
  spec.add_development_dependency 'rubocop-rails_config', '~> 1.8'
  spec.add_development_dependency 'simplecov', '~> 0.8'
  spec.add_development_dependency 'webmock', '~> 3.0'
  spec.add_development_dependency 'yard-tomdoc', '~> 0.7'
```

Among other possible problems, this appears to be causing issues with Dependabot's updates. Presumably, it will also cause non-production/runtime dependencies to be installed in production-only environments (like servers).

The above samples were collected using:

```
% ruby --version
ruby 3.2.0 (2022-12-25 revision a528908271) [x86_64-darwin22]
% gem --version
3.4.3
```

**Comment by tiagobabo** [comment]
We are facing the same issue too. GitHub Registry server is returning development dependencies in the metadata when it shouldn't.

**Comment by helderasmoreira** [comment]
Same here, +1!

**Comment by nbibler** [comment]
~It looks like the `https://x-access-token:ghp_XXXXX@rubygems.pkg.github.co/myorg/specs.4.8.gz` response from the GitHub Private Registry server is incorrect. Putting together a sample to post here.~

I'm not sure that it's the spec.4.8.gz response anymore. That response appears to be correct. I've also downloaded a GitHub-hosted gem and extracted it. The contained `metadata.gz` appears to correctly indicate the `type: :runtime` vs `type: :development` for the listed dependencies (this would make sense since it was built and uploaded externally to GitHub). So, seemingly some process or function within the Registry application which should be interpreting this and providing API responses is likely behaving incorrectly.

**Reply by tiagobabo** [reply]
Yes, it looks like it is a problem with the way GitHub is processing the uploaded gems and creating the metadata. I also tried to download the gem and the `metada.gz` looks good.

**Comment by JoeSouthan** [comment]
Same here, can confirm that unpacking the gem and inspecting the `metadata` everything seems correct but the compact index returned from the package api is including development dependencies.

**Comment by nickarora** [comment]
Confirming we are seeing the same issue.  Here are our observations:

1. The issue affects packages with a gemspec that contains the `add_development_dependency` directive:

```ruby
Gem::Specification.new do |s|
  s.name = "some-package"
  s.version = "1.0.0"
  # ...
  s.add_runtime_dependency "some-dependency"
  s.add_development_dependency "some-development-dependency"
end
```

2. The published gem includes the correct metadata.  After downloading the gem from GH, decompressing the gem, and inspecting the metadata, we see that the development dependency is correctly tagged as `type: :development`

```yaml
- !ruby/object:Gem::Dependency
  name: some-development-dependency
  requirement: !ruby/object:Gem::Requirement
    requirements:
    - - ">="
      - !ruby/object:Gem::Version
        version: '0'
  type: :development
```

3. However, when downloading the Marshal'd gemspec index from `quick/Marshal.4.8/some-package-1.0.0.gemspec.rz`, we see a different story.  In the indexed gemspec, the development dependency is now *incorrectly* tagged *as a runtime dependency.*

```ruby
Gem::Specification.new do |s|
  s.name = "some-package"
  s.version = Gem::Version.new("1.0.0")
  # ...
  s.dependencies = [
    Gem::Dependency.new("some-dependency", Gem::Requirement.new([">= 0"]), :runtime),
    Gem::Dependency.new("some-development-dependency", Gem::Requirement.new([">= 0"]), :runtime)  # <= This is incorrect
  ]
  # ...
end
```

The issue appears to be related to how the index is generated.  However, its hard to say more because GH's package repository implementation is opaque. The issue is problematic because installation of the gem in operational environments pulls in development dependencies.

Hopefully we can see a resolution soon.  Thank you!

**Reply by nbibler** [reply]
Nice job, @nickarora! I stopped short of checking down the `quick/Marshal.4.8/…` requests and apparently should've kept driving down that path. Agreed that there's not much more information we can provide as that's all server-side generated and going to be unique to GitHub's implementation. Thanks for the info!

**Comment by ankitkaushal01** [comment]
@nbibler we have identified the issue & working on the fix. Will update here once the fix is deployed

**Reply by ankitkaushal01** [reply]
We have deployed the fix for the above issue, please verify once.

**Reply by nbibler** [reply]
Thanks for the update @ankitkaushal01 and the fix. I haven't yet had a chance to try it out, but judging by others' responses, there appears to be a new problem (malformed version numbers?).

**Reply by nbibler** [reply]
@ankitkaushal01: The Dependabot issue appears to be corrected. PRs are now correctly modifying the Gemfile.lock with expected changes and the associated GitHub Actions test runs appear to be executing properly for my environments. Thank you! 🍻

I haven't yet experienced the `Malformed version number string` reported by others.

**Reply by nbibler** [reply]
While it looks like Dependabot was fixed, it doesn't look like the actual server responses have been corrected yet. And that's causing the Rubygems CLI (`gem`) to have issues. See https://github.com/orgs/community/discussions/43861#discussioncomment-4659249.

**Comment by andreas-venturini** [comment]
@ankitkaushal01 we're now running into the following errors (already during normal bundling, which used to work before)

```
Retrying fetcher due to error (2/4): ArgumentError Malformed version number string 0.1.2
```

We started seeing those within the last hour or so, this has not happened prior.

**Reply by nbibler** [reply]
I wonder if it's a coincidence that all of the reported `Malformed version number` errors look to be `0.x` releases. I haven't yet seen this error and I don't think I currently have any `0.x` private gems…

**Reply by BytewaveMLP** [reply]
We have a few 0.x private gems, but it appears to be choking on one gem in particular.

**Reply by jdashton** [reply]
> I wonder if it's a coincidence that all of the reported `Malformed version number` errors look to be `0.x` releases. I haven't yet seen this error and I don't think I currently have any `0.x` private gems…

It's coincidence. I'm seeing it with version 1.0.2:
`Malformed version number string 1.0.2|checksum:e87a9`...

**Comment by joshvince** [comment]
Seeing this as well with a different version number

```
Retrying fetcher due to error (2/4): ArgumentError Malformed version number string 0.0.17
```

**Comment by jedrekdomanski** [comment]
Having the same issue
```
Retrying fetcher due to error (2/4): ArgumentError Malformed version number string 0.42.0.pre.b1e6081
```

**Comment by andreas-venturini** [comment]
@ankitkaushal01 would it be possible to revert the fix so at least normal bundling works again? As it is right now our entire devops pipeline is broken including deploys etc.

**Reply by ankitkaushal01** [reply]
is it possible to run in verbose mode & share the logs ? I also tried bundle install but it worked for me.

**Reply by dweinand** [reply]
@ankitkaushal01 Can we please roll back this change before we try to debug this issue? It's causing major issues across the board.

**Reply by nbibler** [reply]
I've marked this Discussion as **Answered** even though it looks like there's likely some other problem now arising. I didn't mark it Answered to claim it was complete, but that the focus of the discussion appears to have been addressed – albeit with a new problem introduced.

**Reply by ankitkaushal01** [reply]
@andreas-venturini can you please check now & verify it once ?

**Reply by BytewaveMLP** [reply]
@ankitkaushal01 Seems to be working now!

**Reply by dweinand** [reply]
@ankitkaushal01 We're now getting gem version conflicts with development dependencies of our published gems and our local app.

**Reply by andreas-venturini** [reply]
@ankitkaushal01 I can confirm that bundling works again, thanks for reverting.

@dweinand this is the original issue again (see top post) - so this still needs to be fixed

**Reply by ankitkaushal01** [reply]
@dweinand can you please try again & confirm is this issue "gem version conflicts with development dependencies" still coming ?

**Comment by nickarora** [comment]
**Edit: Note I saw this before the above reversion.  But keeping it here for posterity.**
**Edit2: The problem is back**

@ankitkaushal01 As a result of the fix, I am seeing a new regression:  Installing gems with the `--development` flag does not install dependencies specified using the `add_development_dependency` directive.

Here's what I'm observing -- and why I suspect it is happening:

For a package with a gemspec that contains the add_development_dependency directive:
```ruby
Gem::Specification.new do |s|
  s.name = "some-package"
  s.version = "1.0.0"
  # ...
  s.add_runtime_dependency "some-dependency"
  s.add_development_dependency "some-development-dependency"
end
```

The published gem will include the correct metadata. After downloading the gem from GH, decompressing the gem, and inspecting the metadata, we can see that the development dependency is correctly tagged as type: 
`:development`

```yaml
- !ruby/object:Gem::Dependency
  name: some-development-dependency
  requirement: !ruby/object:Gem::Requirement
    requirements:
    - - ">="
      - !ruby/object:Gem::Version
        version: '0'
  type: :development
```

However, when downloading the Marshal'd gemspec index from quick/Marshal.4.8/some-package-1.0.0.gemspec.rz, I see incorrect output.

What I see:

```ruby
Gem::Specification.new do |s|
  s.name = "some-package"
  s.version = Gem::Version.new("1.0.0")
  # ...
  s.dependencies = [
    Gem::Dependency.new("some-dependency", Gem::Requirement.new([">= 0"]), :runtime),
  ]
  # ...
end
```

What I _should_ see:
```ruby
Gem::Specification.new do |s|
  s.name = "some-package"
  s.version = Gem::Version.new("1.0.0")
  # ...
  s.dependencies = [
    Gem::Dependency.new("some-dependency", Gem::Requirement.new([">= 0"]), :runtime),
    Gem::Dependency.new("some-development-dependency", Gem::Requirement.new([">= 0"]), :development)
  ]
  # ...
end
```

The omission of the development dependency affects our ability to install using the `--development` flag when using [gem install](https://guides.rubygems.org/command-reference/#gem-install).

@nbibler are you seeing similar behavior?

**Reply by nbibler** [reply]
@nickarora : Yes, I'm seeing the same. The `gemspec` returned from the GitHub Registry does indeed _**incorrectly**_ identify the "my-private-gem2" dependency as a `:runtime` dependency rather than a `:development` dependency.

```
curl "https://x-access-token:ghp_XXXXX@rubygems.pkg.github.com/myorg/quick/Marshal.4.8/my-private-gem1-3.0.0.gemspec.rz" -o my-private-gem1-3.0.0.gemspec.rz 
```

```ruby
spec = Marshal.load(Gem::Util.inflate(File.read('my-private-gem1-3.0.0.gemspec.rz')))
# => 
# Gem::Specification.new do |s|                                                                                                      
#   s.name = "my-private-gem1"                                                                                                         
#   s.version = Gem::Version.new("3.0.0")                                                                                            
#   s.installed_by_version = Gem::Version.new("0")                                                                                   
#   s.date = Time.utc(2023, 1, 11)                                                                                                   
#   s.dependencies = [Gem::Dependency.new("addressable", Gem::Requirement.new(["~> 2.3"]), :runtime),                                
#    Gem::Dependency.new("my-private-gem2", Gem::Requirement.new(["~> 2.1"]), :runtime),                                                 
#    Gem::Dependency.new("omniauth", Gem::Requirement.new(["~> 1.1"]), :runtime)]                                                    
#   s.description = "XXXXX"                                        
#   s.require_paths = ["lib"]                                                                                                        
#   s.rubygems_version = "3.1.2"
#   s.specification_version = 4 
#   s.summary = ""              
#   end                   
```

And the Rubygem CLI still incorrectly identifies the dependency types:

```
% gem dependency -r my-private-gem1 -v 3.3.0 -s
"https://x-access-token:ghp_XXXXX@rubygems.pkg.github.com/myorg"
Gem my-private-gem1-3.3.0
  addressable (~> 2.3)
  my-private-gem2 (~> 2.1)
  appraisal (~> 2.0)
  bundler (~> 2.0)
  faraday (~> 2.0)
  license_finder (~> 7.0)
  omniauth (>= 1.1, < 3)
  rack-test (~> 2.0)
  rake (~> 13.0)
  redcarpet (~> 3.1)
  rspec (~> 3.0)
  rubocop-rails_config (~> 1.8)
  simplecov (~> 0.8)
  webmock (~> 3.0)
  yard-tomdoc (~> 0.7)
```

**Reply by nickarora** [reply]
@nbibler 

I believe that you tested after @ankitkaushal01 reverted his changes.  You are seeing the original reported issue again.  

After the (now reverted) fix was pushed, there was a slightly different error. The development dependencies were entirely omitted from the indexed gemspec.  

Hopefully when it's finally resolved, all will work as expected 🤞

**Reply by nbibler** [reply]
😆 Agreed on 🤞!

And you're right. I missed that your example above had the `:development` dependency entirely missing from the response. I read through it too quickly! ☹️

**Comment by ankitkaushal01** [comment, marked as answer]
are you folks still facing "Malformed version number" & "gem version conflicts with development dependencies" these issues ?
@dweinand @andreas-venturini @jedrekdomanski @joshvince

**Reply by andreas-venturini** [reply]
@ankitkaushal01 it's back at the "Malformed version number string" ArgumentError now. Seems like we go round in circles.

**Reply by wayne311** [reply]
I work with @jedrekdomanski , we am still getting the Malformed version number error, the version number has changed.
Malformed version number string 0.1.50

**Reply by nickarora** [reply]
@ankitkaushal01 

I see neither of those problems but the problem I [describe here](https://github.com/orgs/community/discussions/43861#discussioncomment-4659249) is now happening. Installing gems with the `--development` flag is not being respected.  This is likely because the Marshal'd index is being built incorrectly as I describe in my linked comment -- but there may be another underlying issue.

**Reply by jdashton** [reply]
> are you folks still facing "Malformed version number" & "gem version conflicts with development dependencies" these issues ? @dweinand @andreas-venturini @jedrekdomanski @joshvince

The Malformed version number error is gone (again) as of 2:41 PM Eastern Time (New York time).

**Reply by ankitkaushal01** [reply]
Both the issues should be fixed now !!

**Reply by nickarora** [reply]
@ankitkaushal01 The issue I described is still present. Installing a gem with `gem install --development` does not install development dependencies specified using `add_development_dependency`

**Reply by ankitkaushal01** [reply]
@nickarora There is no support right now to install development dependency, we are tracking this request as a separate item & will work on the fix.

**Reply by nickarora** [reply]
@ankitkaushal01 thanks; will you comment here when it is resolved?
